File: apps/api/app/schemas/signals.py
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, Field

from app.models.enums import AssetClass, Bias, ScoreClass, SignalStatus, StrategyType, Timeframe


class SignalReviewEventRead(BaseModel):
    id: int
    signal_id: int
    user_id: int
    event_type: str
    previous_status: SignalStatus | None
    new_status: SignalStatus | None
    note: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


class SignalRead(BaseModel):
    id: int
    watchlist_item_id: int
    symbol: str
    asset_class: AssetClass
    exchange: str | None
    strategy_type: StrategyType
    status: SignalStatus
    bias: Bias
    score: int | None
    score_class: ScoreClass | None
    timeframe_context: Timeframe | None
    timeframe_setup: Timeframe | None
    timeframe_trigger: Timeframe | None
    entry_low: Decimal | None
    entry_high: Decimal | None
    trigger_level: Decimal | None
    stop_loss: Decimal | None
    target_1: Decimal | None
    target_2: Decimal | None
    risk_reward: Decimal | None
    invalidation_reason: str | None
    reasoning: list | dict | None
    risk_flags: list | dict | None
    no_trade_reasons: list | dict | None
    next_action: str | None
    quality_report: list[dict[str, str]] = Field(default_factory=list)
    review_note: str | None
    created_at: datetime
    updated_at: datetime
    triggered_at: datetime | None
    is_stale: bool
    stale_reason: str | None
    stale_after_days: int
    review_events: list[SignalReviewEventRead] = Field(default_factory=list)

    model_config = {"from_attributes": True}


class SignalStatusUpdate(BaseModel):
    status: SignalStatus


class SignalReviewNoteUpdate(BaseModel):
    review_note: str | None = Field(default=None, max_length=5000)

File: apps/web/src/types/signals.ts
export type SignalStatus = "watchlist" | "armed" | "triggered" | "invalidated" | "no_setup" | "missed" | "expired";

export type ScoreClass = "a_setup" | "b_setup" | "watchlist" | "no_trade";

export type Bias = "bullish" | "neutral" | "bearish";

export type StrategyType = "trend_pullback_long" | "base_breakout_long";

export type Timeframe = "1W" | "1D" | "4H";

export type AssetClass = "stock" | "crypto";

export type Signal = {
  id: number;
  watchlist_item_id: number;
  symbol: string;
  asset_class: AssetClass;
  exchange: string | null;
  strategy_type: StrategyType;
  status: SignalStatus;
  bias: Bias;
  score: number | null;
  score_class: ScoreClass | null;
  timeframe_context: Timeframe | null;
  timeframe_setup: Timeframe | null;
  timeframe_trigger: Timeframe | null;
  entry_low: string | null;
  entry_high: string | null;
  trigger_level: string | null;
  stop_loss: string | null;
  target_1: string | null;
  target_2: string | null;
  risk_reward: string | null;
  invalidation_reason: string | null;
  reasoning: string[] | Record<string, unknown> | null;
  risk_flags: string[] | Record<string, unknown> | null;
  no_trade_reasons: string[] | Record<string, unknown> | null;
  next_action: string | null;
  review_note: string | null;
  created_at: string;
  updated_at: string;
  triggered_at: string | null;
};

export type SignalStatusUpdatePayload = {
  status: SignalStatus;
};

export type SignalReviewNoteUpdatePayload = {
  review_note: string | null;
};
